ccm_web/client/src/components/MultipleSectionEnrollmentWorkflow.tsx

```diff
@@ -0,0 +1,341 @@
+  },
+  buttonGroup: {
+    marginTop: theme.spacing(1)
+  },
+  sectionIdTable: {
+    maxHeight: 300
+  },
+  confirmationTable: {
+    paddingRight: theme.spacing(1),
+    paddingLeft: theme.spacing(1)
+  },
+  container: {
+    position: 'relative',
+    zIndex: 0
+  },
+  backdrop: {
+    zIndex: theme.zIndex.drawer + 1,
+    color: '#FFF',
+    position: 'absolute',
+    textAlign: 'center'
+  }
+}))
+
+interface MultipleSectionEnrollmentWorkflowProps extends AddUMUsersLeafProps {
+  course: CanvasCourseBase
+}
+
+export default function MultipleSectionEnrollmentWorkflow (props: MultipleSectionEnrollmentWorkflowProps): JSX.Element {
+  const parser = new FileParserWrapper()
+  const sectionIds = props.sections.map(s => s.id)
+
+  const classes = useStyles()
+  const [workflowState, setWorkflowState] = useState<CSVWorkflowState>(CSVWorkflowState.Upload)
+  const [file, setFile] = useState<File | undefined>(undefined)
+  const [validEnrollments, setValidEnrollments] = useState<RowNumberedAddEnrollmentWithSectionId[] | undefined>(undefined)
+
+  const [schemaInvalidations, setSchemaInvalidations] = useState<SchemaInvalidation[] | undefined>(undefined)
+  const [rowInvalidations, setRowInvalidations] = useState<EnrollmentInvalidation[] | undefined>(undefined)
+
+  const [doAddEnrollments, isAddEnrollmentsLoading, addEnrollmentsError, clearAddEnrollmentsError] = usePromise(
+    async (courseId: number, enrollments: AddEnrollmentWithSectionId[]) => {
+      await api.addEnrollmentsToSections(courseId, enrollments.map(e => ({
+        loginId: e.loginId, type: getCanvasRole(e.role), sectionId: e.sectionId
+      })))
+    },
+    () => setWorkflowState(CSVWorkflowState.Confirmation)
+  )
+
+  const getSectionsErrorAlert = (
+    <ErrorAlert
+      messages={[<Typography key={0}>An error occurred while loading section data from Canvas.</Typography>]}
+      tryAgain={props.doGetSections}
+    />
+  )
+
+  const resetUpload = async (): Promise<void> => {
+    setSchemaInvalidations(undefined)
+    setRowInvalidations(undefined)
+    setValidEnrollments(undefined)
+    clearAddEnrollmentsError()
+    setWorkflowState(CSVWorkflowState.Upload)
+    await props.doGetSections()
+  }
+
+  const renderRowValidationErrors = (errors: RowValidationError[]): JSX.Element => {
+    return (
+      <>
+      {file !== undefined && <CSVFileName file={file} />}
+      <RowLevelErrorsContent
+        table={<ValidationErrorTable invalidations={errors} />}
+        title='Review your CSV file'
+        resetUpload={resetUpload}
+      />
+      </>
+    )
+  }
+
+  const renderSchemaInvalidations = (invalidations: SchemaInvalidation[]): JSX.Element => {
+    const errors = invalidations.map(
+      (invalidation, i) => <Typography key={i}>{invalidation.message}</Typography>
+    )
+    return (
+      <>
+      {file !== undefined && <CSVFileName file={file} />}
+      <ErrorAlert messages={errors} tryAgain={resetUpload} />
+      </>
+    )
+  }
```

Comment:
This is mostly just repeated between CSV features. It would be nice to extract these at some point.

Reply:
#315